docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,569 @@
+
+// KubeadmControlPlane is the Schema for the KubeadmControlPlane API.
+type KubeadmControlPlane struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   KubeadmControlPlaneSpec   `json:"spec,omitempty"`
+    Status KubeadmControlPlaneStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// KubeadmControlPlaneList contains a list of KubeadmControlPlane.
+type KubeadmControlPlaneList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []KubeadmControlPlane `json:"items"`
+}
+
+func init() {
+    SchemeBuilder.Register(&KubeadmControlPlane{}, &KubeadmControlPlaneList{})
+}
+
+type KubeadmControlPlaneStatusError string
+
+// A more descriptive kind of error that represents an error condition that
+// should be set in the KubeadmControlPlane.Status. The "Reason" field is meant for short,
+// enum-style constants meant to be interpreted by control planes. The "Message"
+// field is meant to be read by humans.
+type KubeadmControlPlaneError struct {
+    Reason  KubeadmControlPlaneStatusError
+    Message string
+}
+
+func (e *KubeadmControlPlaneError) Error() string {
+    return e.Message
+}
+
+const (
+    // InvalidConfigurationKubeadmControlPlaneError indicates that the kubeadm control plane
+    // configuration is invalid.
+    InvalidConfigurationKubeadmControlPlaneError KubeadmControlPlaneStatusError = "InvalidConfiguration"
+
+    // UnsupportedChangeKubeadmControlPlaneError indicates that the kubeadm control plane
+    // spec has been updated in an unsupported way. That cannot be
+    // reconciled.
+    UnsupportedChangeKubeadmControlPlaneError KubeadmControlPlaneStatusError = "UnsupportedChange"
+
+    // CreateKubeadmControlPlaneError indicates that an error was encountered
+    // when trying to create the kubeadm control plane.
+    CreateKubeadmControlPlaneError KubeadmControlPlaneStatusError = "CreateError"
+
+    // UpdateKubeadmControlPlaneError indicates that an error was encountered
+    // when trying to update the kubeadm control plane.
+    UpdateKubeadmControlPlaneError KubeadmControlPlaneStatusError = "UpdateError"
+
+    // DeleteKubeadmControlPlaneError indicates that an error was encountered
+    // when trying to delete the kubeadm control plane.
+    DeleteKubeadmControlPlaneError KubeadmControlPlaneStatusError = "DeleteError"
+)
+```
+
+With the following validations:
+
+- If `KubeadmControlPlane.Spec.KubeadmConfigSpec` does not define external etcd (webhook):
+  - `KubeadmControlPlane.Spec.Replicas` is an odd number.
+  - Configuration of external etcd is determined by introspecting the provided `KubeadmConfigSpec`.
+- `KubeadmControlPlane.Spec.Replicas` is >= 0 or is nil
+- `KubeadmControlPlane.Spec.Version != ""` (openapi)
+- `KubeadmControlPlane.Spec.KubeadmConfigSpec` must be treated as immutable (via webhook)
```

Comment:
Would we want to allow modifications to `Pre/PostKubeadmCommands`, `Files`?

Reply:
Do you have specific use cases in mind? It might get a bit tricky to determine what are allowed vs not allowed modifications. We already capture the lack of mutability here in the Risks section, since once we adopt v1beta2 types for kubeadm there may need to be changes made to the kustomize templates provided.

Reply:
Nothing specifc. Was trying to reconcile what it said below about triggering an update based on changes to this. I'm happy to start with it immutable, and loosen if we have feedback that it's needed.